tidy3d/plugins/microwave/custom_path_integrals.py

```diff
@@ -0,0 +1,163 @@
+        The :class:``FieldParameter`` determines whether the electric or magnetic field is integrated.
+        """
+        if not isinstance(em_field, (FieldData, FieldTimeData, ModeSolverData)):
+            raise DataError("'em_field' type not supported.")
+
+        (dim1, dim2, dim3) = self.local_dims
+
+        h_field_name = f"{field}{dim1}"
+        v_field_name = f"{field}{dim2}"
+        # Validate that fields are present
+        if h_field_name not in em_field.field_components:
+            raise DataError(f"'field_name' '{h_field_name}' not found.")
+        if v_field_name not in em_field.field_components:
+            raise DataError(f"'field_name' '{v_field_name}' not found.")
+
+        # Select fields lying on the plane
+        plane_indexer = {dim3: self.position}
+        field1 = em_field.field_components[h_field_name].sel(plane_indexer, method="nearest")
+        field2 = em_field.field_components[v_field_name].sel(plane_indexer, method="nearest")
+
+        # Dim 's' represents the parameterization of the line
+        # 't' is likely used for time
+        x_path = xr.DataArray(self.vertices[:, 0], dims="s")
+        y_path = xr.DataArray(self.vertices[:, 1], dims="s")
+        field1_interp = field1.interp(x=x_path, y=y_path, method="linear")
+        field2_interp = field2.interp(x=x_path, y=y_path, method="linear")
+
+        # Determine the differential length elements along the path
+        dl_x = self._compute_dl_component(x_path, self.is_closed_contour)
+        dl_y = self._compute_dl_component(y_path, self.is_closed_contour)
+        dl_x = xr.DataArray(dl_x, dims="s")
+        dl_y = xr.DataArray(dl_y, dims="s")
+
+        # Compute the dot product between differential length element and vector field
+        integrand = field1_interp * dl_x + field2_interp * dl_y
+        # Integrate along the path
+        result = integrand.integrate(coord="s")
+        result = result.reset_coords(drop=True)
+
+        if isinstance(em_field, FieldData):
+            return FreqDataArray(data=result.data, coords=result.coords)
+        elif isinstance(em_field, FieldTimeData):
+            return TimeDataArray(data=result.data, coords=result.coords)
+        else:
+            assert isinstance(em_field, ModeSolverData)
+            return FreqModeDataArray(data=result.data, coords=result.coords)
+
+    @staticmethod
+    def _compute_dl_component(coord_array: xr.DataArray, closed_contour=False) -> np.array:
+        """Computes the differential length element along the integration path."""
+        dl = np.gradient(coord_array)
+        if closed_contour:
+            # If the contour is closed, we can use central different on the starting/end point
+            # which will be more accurate than the default forward/backward choice in np.gradient
+            grad_end = np.gradient([coord_array[-2], coord_array[0], coord_array[1]])
+            dl[0] = dl[-1] = grad_end[1]
```

Comment:
This is equivalent to `dl[0] = dl[-1] = (coord_array[1]-coord_array[-2])/2`? If so, this is clearer and more efficient than `np.gradient`?

Reply:
@tylerflex made the opposite point :smiley:  I think reusing the gradient function is also pretty clear, it shows that we are applying the same operation but just in a periodic fashion. Your suggestion is probably more efficient, but maybe not observable.

Reply:
I dont think it's equivalent @weiliangjin2021 . The `np.gradient` I think uses a 2nd order derivative, [link](https://numpy.org/doc/stable/reference/generated/numpy.gradient.html). and since the rest of this function uses `np.gradient()` I assumed it might make sense to use the same scheme over the whole array.

Reply:
In the case where the spacing is uniform it should be equivalent, which should correspond to my case. See the last equation in that link.

Reply:
The geometric spacing between vertices can already be non-uniform. The numpy gradient function is just a more generalized function where you could give it coordinates to perform the gradient over. We are simply using it to perform central differences, where the implied spacing is 1.

Reply:
ooh, I see.

Reply:
another reason why I like np.gradient: I think I mentioned in another comment, that I prefer the more declarative, high level stuff (like `np.gradient()`, `pop_axis()` all else equal. And if we do have to perform mathematical operations, such as `(f1-f2)/2.0`, ideally that stuff should be abstracted away in its own function. I think this helps with readability.

Reply:
I forgot about nonuniform vertices. Then np.gradient is a good choice here.

Reply:
After playing around with a notebook, using some of these computations more, I think I will change the behaviour of custom path integral. Think about a rectangular path where the user has defined path `vertices` as the corners of a rectangle. The computed `dl` is not accurate at the corner. It is probably better to interpolate the fields at the midpoints of each line segment, where the `dl` can be computed accurately.

Reply:
Now, after sleeping on it. I decided to keep it the way it is. As long as enough vertices are used the result is accurate, and what I mentioned above would be much more complicated to implement.